// stryker/src/SandboxPool.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var os = require("os");
var log4js_1 = require("log4js");
var rxjs_1 = require("rxjs");
var Sandbox_1 = require("./Sandbox");
var SandboxPool = /** @class */ (function () {
    function SandboxPool(options, testFramework, initialFiles) {
        this.options = options;
        this.testFramework = testFramework;
        this.initialFiles = initialFiles;
        this.log = log4js_1.getLogger(SandboxPool.name);
        this.sandboxes = [];
        this.isDisposed = false;
    }
    SandboxPool.prototype.streamSandboxes = function () {
        var _this = this;
        var numConcurrentRunners = os.cpus().length;
        if (this.options.transpilers.length) {
            // If transpilers are configured, one core is reserved for the compiler (for now)
            numConcurrentRunners--;
        }
        var numConcurrentRunnersSource = 'CPU count';
        if (numConcurrentRunners > this.options.maxConcurrentTestRunners && this.options.maxConcurrentTestRunners > 0) {
            numConcurrentRunners = this.options.maxConcurrentTestRunners;
            numConcurrentRunnersSource = 'maxConcurrentTestRunners config';
        }
        if (numConcurrentRunners <= 0) {
            numConcurrentRunners = 1;
        }
        this.log.info("Creating " + numConcurrentRunners + " test runners (based on " + numConcurrentRunnersSource + ")");
        var sandboxes = rxjs_1.Observable.range(0, numConcurrentRunners)
            .flatMap(function (n) { return _this.registerSandbox(Sandbox_1.default.create(_this.options, n, _this.initialFiles, _this.testFramework)); });
        return sandboxes;
    };
    SandboxPool.prototype.registerSandbox = function (promisedSandbox) {
        var _this = this;
        return promisedSandbox.then(function (sandbox) {
            if (_this.isDisposed) {
                // This sandbox is too late for the party. Dispose it to prevent hanging child processes
                // See issue #396
                sandbox.dispose();
            }
            else {
                _this.sandboxes.push(sandbox);
            }
            return sandbox;
        });
    };
    SandboxPool.prototype.disposeAll = function () {
        this.isDisposed = true;
        return Promise.all(this.sandboxes.map(function (sandbox) { return sandbox.dispose(); }));
    };
    return SandboxPool;
}());
exports.default = SandboxPool;
